File: internal/plugins/typescript/rules/prefer_includes/prefer_includes_test.go
package prefer_includes

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestPreferIncludesRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &PreferIncludesRule, []rule_tester.ValidTestCase{
		// indexOf without comparison
		{Code: "foo.indexOf(bar)"},
		{Code: "foo.indexOf(bar) + 0"},

		// Union types
		{Code: `
declare const val: string | number;
val.indexOf('foo') === -1;
		`},

		// User-defined types without includes
		{Code: `
type UserType = { indexOf(x: string): number };
declare const foo: UserType;
foo.indexOf('x') === -1;
		`},

		// User-defined types with different includes signature
		{Code: `
type WithIncludesButDifferent = {
	indexOf(x: string): number;
	includes(x: string, fromIndex: number): boolean;
};
declare const foo: WithIncludesButDifferent;
foo.indexOf('x') === -1;
		`},

		// includes is a boolean, not a method
		{Code: `
type WithBooleanIncludes = {
	indexOf(x: string): number;
	includes: boolean;
};
declare const foo: WithBooleanIncludes;
foo.indexOf('x') === -1;
		`},

		// RegExp with flags
		{Code: `/bar/i.test(foo)`},

		// RegExp with character class
		{Code: `/ba[rz]/.test(foo)`},

		// RegExp with alternation
		{Code: `/foo|bar/.test(foo)`},

		// RegExp test without arguments
		{Code: `pattern.test()`},

		// Non-RegExp test method
		{Code: `
type WithTest = { test(x: string): boolean };
declare const obj: WithTest;
obj.test(foo);
		`},

		// RegExp from variable
		{Code: `
const pattern = /foo/;
const regex = pattern;
regex.test(str);
		`},
	}, []rule_tester.InvalidTestCase{
		// Basic indexOf comparisons - positive checks (should use includes)
		{
			Code: `foo.indexOf(bar) !== -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) != -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) > -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) >= 0`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},

		// Basic indexOf comparisons - negative checks (should use !includes)
		{
			Code: `foo.indexOf(bar) === -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) == -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) < 0`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},
		{
			Code: `foo.indexOf(bar) <= -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},

		// Reversed operands
		{
			Code: `-1 !== foo.indexOf(bar)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},
		{
			Code: `0 <= foo.indexOf(bar)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes(bar)`},
		},
		{
			Code: `-1 === foo.indexOf(bar)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},
		{
			Code: `0 > foo.indexOf(bar)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
			Output: []string{`!foo.includes(bar)`},
		},

		// Optional chaining - should report but not fix
		{
			Code: `a?.indexOf(b) === -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
		},
		{
			Code: `a?.indexOf(b) !== -1`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 1, Column: 1},
			},
		},

		// RegExp.test() cases
		{
			Code: `/bar/.test(foo)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferStringIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes("bar")`},
		},
		{
			Code: `/bar/.test((1 + 1, foo))`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferStringIncludes", Line: 1, Column: 1},
			},
			Output: []string{`(1 + 1, foo).includes("bar")`},
		},
		{
			Code: `/\\0'\\\\\\n\\r\\v\\t\\f/.test(foo)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferStringIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes("\\0'\\\\\\n\\r\\v\\t\\f")`},
		},
		{
			Code: `new RegExp('bar').test(foo)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferStringIncludes", Line: 1, Column: 1},
			},
			Output: []string{`foo.includes("bar")`},
		},
		{
			Code: `
const pattern = 'bar';
new RegExp(pattern).test(foo + bar);
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferStringIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
const pattern = 'bar';
(foo + bar).includes("bar");
			`},
		},

		// Array types
		{
			Code: `
declare const arr: any[];
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: any[];
arr.includes(x);
			`},
		},
		{
			Code: `
declare const arr: ReadonlyArray<any>;
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: ReadonlyArray<any>;
arr.includes(x);
			`},
		},

		// TypedArrays
		{
			Code: `
declare const arr: Int8Array;
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: Int8Array;
arr.includes(x);
			`},
		},
		{
			Code: `
declare const arr: Uint8Array;
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: Uint8Array;
arr.includes(x);
			`},
		},
		{
			Code: `
declare const arr: Float32Array;
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: Float32Array;
arr.includes(x);
			`},
		},

		// Generic constraints
		{
			Code: `
function fn<T extends string>(x: T) {
	return x.indexOf('a') !== -1;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 9},
			},
			Output: []string{`
function fn<T extends string>(x: T) {
	return x.includes('a');
}
			`},
		},

		// Readonly arrays
		{
			Code: `
declare const arr: Readonly<any[]>;
arr.indexOf(x) !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 3, Column: 1},
			},
			Output: []string{`
declare const arr: Readonly<any[]>;
arr.includes(x);
			`},
		},

		// User-defined types with matching indexOf and includes
		{
			Code: `
type WithBothMethods = {
	indexOf(x: string): number;
	includes(x: string): boolean;
};
declare const foo: WithBothMethods;
foo.indexOf('x') !== -1;
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "preferIncludes", Line: 7, Column: 1},
			},
			Output: []string{`
type WithBothMethods = {
	indexOf(x: string): number;
	includes(x: string): boolean;
};
declare const foo: WithBothMethods;
foo.includes('x');
			`},
		},
	})
}

File: packages/rslint-test-tools/tests/typescript-eslint/rules/prefer-includes.test.ts
import { RuleTester } from '@typescript-eslint/rule-tester';

import { getFixturesRootDir } from '../RuleTester';

const rootDir = getFixturesRootDir();
const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootDir,
    },
  },
});

ruleTester.run('prefer-includes', {
  valid: [
    'foo.indexOf(bar)',
    'foo.indexOf(bar) + 1',
    'foo.indexOf(bar) - 1',
    `
      declare const val: string | number;
      val.indexOf('foo') === -1;
    `,
    `
      type UserDefinedType = { indexOf(x: string): number };
      declare const foo: UserDefinedType;
      foo.indexOf('x') === -1;
    `,
    `
      type WithDifferentIncludes = {
        indexOf(x: string): number;
        includes(x: string, fromIndex: number): boolean;
      };
      declare const foo: WithDifferentIncludes;
      foo.indexOf('x') === -1;
    `,
    `
      type WithBooleanIncludes = {
        indexOf(x: string): number;
        includes: boolean;
      };
      declare const foo: WithBooleanIncludes;
      foo.indexOf('x') === -1;
    `,
    '/bar/i.test(foo)',
    '/ba[rz]/.test(foo)',
    '/foo|bar/.test(foo)',
    'pattern.test()',
    `
      type WithTest = { test(x: string): boolean };
      declare const obj: WithTest;
      obj.test(foo);
    `,
    `
      const pattern = /foo/;
      const regex = pattern;
      regex.test(str);
    `,
  ],
  invalid: [
    {
      code: 'foo.indexOf(bar) !== -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) != -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) > -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) >= 0',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) === -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) == -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) < 0',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: 'foo.indexOf(bar) <= -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: '-1 !== foo.indexOf(bar)',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: '0 <= foo.indexOf(bar)',
      errors: [{ messageId: 'preferIncludes' }],
      output: 'foo.includes(bar)',
    },
    {
      code: '-1 === foo.indexOf(bar)',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: '0 > foo.indexOf(bar)',
      errors: [{ messageId: 'preferIncludes' }],
      output: '!foo.includes(bar)',
    },
    {
      code: 'a?.indexOf(b) === -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: null,
    },
    {
      code: 'a?.indexOf(b) !== -1',
      errors: [{ messageId: 'preferIncludes' }],
      output: null,
    },
    {
      code: '/bar/.test(foo)',
      errors: [{ messageId: 'preferStringIncludes' }],
      output: 'foo.includes("bar")',
    },
    {
      code: '/bar/.test((1 + 1, foo))',
      errors: [{ messageId: 'preferStringIncludes' }],
      output: '(1 + 1, foo).includes("bar")',
    },
    {
      code: String.raw`/\0'\\n\r\v\t\f/.test(foo)`,
      errors: [{ messageId: 'preferStringIncludes' }],
      output: String.raw`foo.includes("\0'\\n\r\v\t\f")`,
    },
    {
      code: "new RegExp('bar').test(foo)",
      errors: [{ messageId: 'preferStringIncludes' }],
      output: 'foo.includes("bar")',
    },
    {
      code: `
        const pattern = 'bar';
        new RegExp(pattern).test(foo + bar);
      `,
      errors: [{ messageId: 'preferStringIncludes' }],
      output: `
        const pattern = 'bar';
        (foo + bar).includes("bar");
      `,
    },
    {
      code: `
        declare const arr: any[];
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: any[];
        arr.includes(x);
      `,
    },
    {
      code: `
        declare const arr: ReadonlyArray<any>;
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: ReadonlyArray<any>;
        arr.includes(x);
      `,
    },
    {
      code: `
        declare const arr: Int8Array;
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: Int8Array;
        arr.includes(x);
      `,
    },
    {
      code: `
        declare const arr: Uint8Array;
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: Uint8Array;
        arr.includes(x);
      `,
    },
    {
      code: `
        declare const arr: Float32Array;
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: Float32Array;
        arr.includes(x);
      `,
    },
    {
      code: `
        function fn<T extends string>(x: T) {
          return x.indexOf('a') !== -1;
        }
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        function fn<T extends string>(x: T) {
          return x.includes('a');
        }
      `,
    },
    {
      code: `
        declare const arr: Readonly<any[]>;
        arr.indexOf(x) !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        declare const arr: Readonly<any[]>;
        arr.includes(x);
      `,
    },
    {
      code: `
        type WithBothMethods = {
          indexOf(x: string): number;
          includes(x: string): boolean;
        };
        declare const foo: WithBothMethods;
        foo.indexOf('x') !== -1;
      `,
      errors: [{ messageId: 'preferIncludes' }],
      output: `
        type WithBothMethods = {
          indexOf(x: string): number;
          includes(x: string): boolean;
        };
        declare const foo: WithBothMethods;
        foo.includes('x');
      `,
    },
  ],
});
